src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs

```diff
@@ -1187,16 +1190,43 @@ internal ImmutableArray<LocalInfo<TypeSymbol>> GetLocalInfo(StandaloneSignatureH
         }
 
         /// <exception cref="UnsupportedSignatureContent">If the encoded parameter type is invalid.</exception>
-        private void DecodeParameterOrThrow(ref BlobReader signatureReader, /*out*/ ref ParamInfo<TypeSymbol> info)
+        private void DecodeParameterOrThrow(ref BlobReader signatureReader, /*out*/ ref ParamInfo<TypeSymbol> info, bool allowOutAttribute = false)
         {
             info.CustomModifiers = DecodeModifiersOrThrow(
                 ref signatureReader,
-                AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute,
+                allowOutAttribute
+                 ? AllowedRequiredModifierType.System_Runtime_InteropServices_InOrOutAttribute
+                 : AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute,
                 out SignatureTypeCode typeCode,
                 out bool inAttributeFound);
 
             if (typeCode == SignatureTypeCode.ByReference)
             {
+                if (allowOutAttribute && !info.CustomModifiers.IsDefaultOrEmpty)
+                {
+                    bool hasOut = false;
+                    bool hasIn = false;
+                    foreach (var modifier in info.CustomModifiers)
```

Comment:
>foreach (var modifier in info.CustomModifiers) [](start = 20, length = 46)

I think this is where approach from [https://github.com/dotnet/roslyn/pull/43275](https://github.com/dotnet/roslyn/pull/43275) would work much better. We would be able to perform the check without having to iterate through modifiers and doing type checks again.